### PR TITLE
Testing under Linux

### PR DESCRIPTION
**Pull request by ns-rse**
* [x] 1 Getting Started with Conda
* [x] 2 Working with Environments
* [x] 3 Using Packages and Channels
* [x] 4 Sharing Environments
* [x] 5 Managing GPU dependencies

### CODE REVIEW

File: _episodes/05-managing-cuda-dependencies.md [hunk 1]
```diff
@@ -219,62 +331,72 @@ channels:
...
-> >   - python=3.7
-> >   - tensorflow-gpu=2.1 # installs tensorflow=2.1=gpu_py37h7a4bb67_0
+> >   - cudatoolkit=11.8
+> >   - cudnn=8.4
+    > >   - cupti=11.3
+> >   - mpi4py=3.1 # installs cuda-aware openmpi
+> >   - nccl=2.14
+> >   - pip=22.3
+> >   - python=3.10
+> >   - tensorflow-gpu=2.11 # installs tensorflow=2.1=gpu_py37h7a4bb67_0
```
Comment (bobturneruk) [inline _episodes/05-managing-cuda-dependencies.md]: Think I can only get 2.6 on Windows.

File: _episodes/02-working-with-environments.md [hunk 2]
```diff
@@ -433,7 +428,7 @@ longer need an existing environment and just want to get rid of cruft on your ma
 the reason, the command to delete an environment is the following.
 
 ~~~
-$ conda remove --name my-first-conda-env --all
+$ conda remove --name first-conda-env --all
```
Comment (bobturneruk) [inline _episodes/02-working-with-environments.md:431]: Have we created an environment called this, by this point i.e. will this code run?

File: _episodes/03-using-packages-and-channels.md [hunk 3]
```diff
@@ -106,25 +106,25 @@ passing the `--channel` option to the `conda install` command as follows.
...
-$ conda install scipy=1.6 --channel conda-forge
+$ conda install scipy=1.10.0 --channel conda-forge
 ~~~
 {: .language-bash}
 
 
 You can also install a package from a specific channel into a named environment (using `--name`). For example, the
 following command installs the `scipy` package from the `conda-forge` channel into the environment
-called `my-first-conda-env` which we created earlier.
+called `first-conda-env` which we created earlier.
```
Comment (bobturneruk) [inline _episodes/03-using-packages-and-channels.md]:
```suggestion
called `machine-learning-env` which we created earlier.

```

File: _episodes/03-using-packages-and-channels.md [hunk 4]
```diff
@@ -349,18 +353,18 @@ $ which pip
...
 > If your `base` Conda environment becomes cluttered with a mix of `pip` and Conda installed
-> packages it may no longer function. Creating separate conda environments allows you to
+> packages it may no longer function. Creating separate Conda environments allows you to
 > delete and recreate environments readily so you dont have to worry about risking your core
 > Conda functionality when mixing packages installed with Conda and Pip.
 {: .callout}
 
 If you find yourself needing to install a Python package that is only available via PyPI, then you should first install
-`pip` into your Conda environment and then use that `pip` to install  the desired package from PyPI. Using the `pip`
+`pip` into your Conda environment and then use that `pip` to install the desired package from PyPI. Using the `pip`
```
Comment (bobturneruk) [inline _episodes/03-using-packages-and-channels.md:329]: Actually, we wouldn't explicitly install `pip` these days as it's free with every Python.

File: _episodes/04-sharing-environments.md [hunk 5]
```diff
@@ -135,27 +144,31 @@ originally created `machine-learning-env` yet from the output of the `conda env
...
 To export this list into an environment.yml file, you can use `--file` option to directly save the
-resulting YAML environment into a file.
+resulting YAML environment into a file. If the target for `--file` exists it will be over-written so make sure your
+filename is unique. So that we do not over-write `environment.yaml` we save the output to `machine-learning-env.yaml`
+instead and add it to
 
 ~~~
-$ conda env export --name machine-learning-env --file environment.yml
+$ conda env export --name machine-learning-env --file machine-learning-env.yml
+$ git init
```
Comment (bobturneruk) [inline _episodes/04-sharing-environments.md:163]: I think it's good to include this much git.

File: _episodes/04-sharing-environments.md [hunk 6]
```diff
@@ -135,27 +144,31 @@ originally created `machine-learning-env` yet from the output of the `conda env
...
-Make sure you do not have any other `environment.yml` file from before in the same directory when
-running the above command. If you do you can use an alternative destination filename, e.g. `--file
-machline-learning-env.yaml`.
 
 This exported environment file may not *consistently* produce environments that are reproducible
 across operating systems. The reason for this is, that it may include operating system specific low-level
 packages which cannot be used by other operating systems.
 
-If you need an environment file that can produce environments that are reproducibile across Mac OS, Windows,
+If you need an environment file that can produce environments that are reproducible across Mac OS, Windows,
```
Comment (bobturneruk) [inline _episodes/04-sharing-environments.md]: I think we need to unpack this and line 165. Maybe not in this PR.

File: _episodes/04-sharing-environments.md [hunk 7]
```diff
@@ -314,18 +340,18 @@ The `--prune` option tells Conda to remove any dependencies that are no longer r
 >  - matplotlib=3.1
 >  - pandas=0.24
 >  - scikit-learn=0.21
->  - pip=19.1
+>  - pip=22.3
```
Comment (bobturneruk) [inline _episodes/04-sharing-environments.md:355]: Maybe Python instead?

File: _episodes/04-sharing-environments.md [hunk 8]
```diff
@@ -344,6 +370,18 @@ The `--prune` option tells Conda to remove any dependencies that are no longer r
 >
 > Conda will then install your `pip` dependencies using `python -m pip install -r requirements.txt`
 > (after creating the Conda environment and installing all Conda installable dependencies).
+>
+> A `requirements.txt` file has a similar structure although it does not use YAML markup, instead it simply lists the
+> packages by name. If a specific version is required then as above it is specified with `==`. Remember you should not
+> include `pip` in the `requirements.txt` because this should be installed by and managed by Conda in your environment.
+>
+> If you use a `requirements.txt` file then you should add this to your Git repository so it too is maintained under
```
Comment (bobturneruk) [inline _episodes/04-sharing-environments.md:390]: Good point!

**Review by bobturneruk — APPROVED** [review]
Looking good. I've skipped over ep5 as I've proposed pretty major changes there. Maybe get this one merged then deal with the conflicts on mine?

File: _episodes/02-working-with-environments.md (hunk 2 above)
Comment (ns-rse) [reply _episodes/02-working-with-environments.md:431]: Good catch, we hadn't. That is what I get for trying to be clever. I've opted to remove `python3-env` which is definitely created but not used anywhere else

File: _episodes/04-sharing-environments.md (hunk 5 above)
Comment (ns-rse) [reply _episodes/04-sharing-environments.md:163]: Cool, enough but no more.

File: _episodes/04-sharing-environments.md (hunk 6 above)
Comment (ns-rse) [reply _episodes/04-sharing-environments.md]: Agree, reading through it now after a gap its a bit brief and could do with more detail.

File: _episodes/04-sharing-environments.md (hunk 7 above)
Comment (ns-rse) [reply _episodes/04-sharing-environments.md:355]: Ah, if `pip` is bundled with Python by default there are quite a few places that need changing to remove explicit mention of `pip` and instead just reference `python`. Back in Episode 3 there is an explanation about avoiding using the system or Conda base `pip`.

On my branch under episode 3 line 364/365 (where lines have been broken so numbers don't align, its the config I've got setup on Emacs and I learnt that Markdown renders them as continuous lines anyway).

> If you find yourself needing to install a Python package that is only available via PyPI, then you should first install
`pip` into your Conda environment and then use that `pip` to install the desired package from PyPI.

Then under the solution `pip` is explicitly installed...

```
> > $ conda install name machine-learning-env pip
```

So there is a bit more work to be done on this front, but as with the above issue I'll save that for another round of reviews and get our current PRs merged.

**Comment by bobturneruk** [comment]
Oh snot. I forgot about the conflicts.

**Pull request merged**